File: src/main/cpp/communication.cpp
/*
 * Copyright © 2023-2024 University of Texas at Arlington
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "tensor.h"
#include <sstream>
#include <cstring>
#include <chrono>
#include <list>
#include <exception>
#include <unistd.h>
#include "opr.h"
#include "comm.h"

extern bool delete_arrays;
extern vector<Opr*> operations;
extern vector<void*(*)(void*)> functions;
extern bool stop_receiver;
extern bool stop_sender;
extern bool inMemory;
extern bool enable_recovery;

int num_of_executors = 1;
int executor_rank = 0;
int coordinator = 0;
int max_buffer_size = 50000000;
const auto comm = MPI_COMM_WORLD;
MPI_Request receive_request = MPI_REQUEST_NULL;
list<int> failed_executors;
list<int> active_executors;
extern bool skip_work;
// max wait time in msecs when sending a message before we assume the receiver is dead
int max_wait_time = 1000;

void info ( const char *fmt, ... );

int serialize ( void* data, char* buffer, vector<int>* encoded_type );
void deserialize ( void* &data, const char* buffer, size_t len, vector<int>* encoded_type );

void delete_block ( void* &data, vector<int>* encoded_type );
void delete_first_reduce_input ( int rid );

void recover ( int failed_executor, int new_executor );

struct {
  MPI_Op acc;
  bool exit;
  bool total;
  int count;
} accumulator;

void reset_accumulator () {
  accumulator.exit = true;
  accumulator.total = true;
  accumulator.count = 0;
}

bool accumulator_exit () { return accumulator.exit; }

bool member ( int n, list<int> v ) {
  for ( auto it = v.begin(); it != v.end(); ++it )
    if (*it == n)
      return true;
  return false;
}

void mpi_error ( int failed_executor,  int error_code ) {
  static char* err_buffer = new char[1000];
  int len;
  err_buffer[0] = 0;
  if (error_code >= 0)
    MPI_Error_string(error_code,err_buffer,&len);
  info("MPI error sending to %d: %s",failed_executor,err_buffer);
  if (enable_recovery && failed_executor >= 0) {
    if (!member(failed_executor,failed_executors)) {
      // send a message to the coordinator that there is a failed executor
      send_long(coordinator,(long)failed_executor,3);
    }
  } else MPI_Abort(comm,-1);
}

void handle_received_message ( int tag, int opr_id, void* data, int len ) {
  if (tag == 0) {
    Opr* opr = operations[opr_id];
    info("    received %d bytes from %d (opr %d)",len,opr->node,opr_id);
    cache_data(opr,data);
    opr->status = completed;
    enqueue_ready_operations(opr_id);
  } else if (tag == 2) {   // cache data
    Opr* opr = operations[opr_id];
    info("    received %d bytes from %d (opr %d)",len,opr->node,opr_id);
    opr->cached = data;
    opr->status = completed;
  } else if (tag == 3) {  // on coordinator only
    // the coordinator will decide who will replace a failed executor
    int failed_executor = (int)((long)data);
    if (!member(failed_executor,failed_executors)) {
      skip_work = true;
      failed_executors.push_back(failed_executor);
      active_executors.remove(failed_executor);
      int new_executor = coordinator;
      for ( int i: active_executors )
        if (i != coordinator)
          new_executor = i;
      if (new_executor < 0) {
        info("Run out of executors - aborting");
        MPI_Abort(comm,-1);
      }
      info("The coordinator has decided to replace the failed executor %d with executor %d",
           failed_executor,new_executor);
      static int* dt = new int[2];
      dt[0] = failed_executor;
      dt[1] = new_executor;
      // announce it to all nodes
      for ( int i: active_executors )
        send_long(i,*(long*)dt,4);
    }
  } else if (tag == 4) {
    // recovery at every executor (sent from coordinator)
    long dt = (long)data;
    uint32_t failed_executor = (uint32_t) dt;
    uint32_t new_executor = (uint32_t) (dt >> 32);
    if (!member(failed_executor,failed_executors)) {
      failed_executors.push_back(failed_executor);
      active_executors.remove(failed_executor);
    }
    recover(failed_executor,new_executor);
  } else if (tag == 5) {
    // synchronize accumulation (on coordinator)
    bool b = (long)data > 0L;
    accumulator.count++;
    accumulator.total = (accumulator.acc == MPI_LOR)
                         ? (accumulator.total || b)
                         : (accumulator.total && b);
    if (!skip_work && accumulator.count == active_executors.size()) {
      long n = accumulator.total ? 1L : 0L;
      for ( int x: active_executors )
        send_long(x,n,6);
    }
  } else if (tag == 6) {
    // release the barrier synchronization after accumulation
    accumulator.exit = true;
    accumulator.count = 0;
    accumulator.total = (long)data > 0L;
  } else if (tag == 1 && operations[opr_id]->type == reduceOPR) {
    Opr* opr = operations[opr_id];
    static tuple<void*,void*>* op_arg = new tuple<void*,void*>(nullptr,nullptr);
    info("    received partial reduce result of %d bytes (opr %d)",
         len,opr_id);
    auto op = (void*(*)(tuple<void*,void*>*))functions[opr->opr.reduce_opr->op];
    if (opr->cached == nullptr) {
      // first reduction input
      info("    set reduce opr %d to the first incoming input",opr_id);
      opr->cached = data;
    } else if (opr->opr.reduce_opr->valuep) {
      // total aggregation
      get<0>(*op_arg) = opr->cached;
      get<1>(*op_arg) = data;
      opr->cached = op(op_arg);
    } else {
      auto x = (tuple<void*,void*>*)opr->cached;
      auto y = (tuple<void*,void*>*)data;
      // merge the current state with the incoming partially reduced data
      info("    merge reduce opr %d with incoming input",opr_id);
      auto old = (void*)x;
      get<0>(*op_arg) = get<1>(*x);
      get<1>(*op_arg) = get<1>(*y);
      opr->cached = new tuple<void*,void*>(get<0>(*x),op(op_arg));
      if (delete_arrays && opr->first_reduced_input < 0) {
        info("    delete current reduce result in opr %d",opr_id);
        delete_block(old,opr->encoded_type);
      }
      if (delete_arrays) {
        info("    delete incoming partial reduce block for opr %d",opr_id);
        delete_block(data,opr->encoded_type);
      }
    }
    delete_first_reduce_input(opr_id);
    opr->reduced_count--;
    if (opr->reduced_count <= 0) {
      // completed final reduce
      info("    completed reduce opr %d",opr_id);
      opr->status = completed;
      enqueue_ready_operations(opr_id);
    }
  }
}

// check if there is a request to send data (array blocks); if there is, get the data and cache it
int check_communication () {
  static char* buffer = new char[max_buffer_size];
  if (receive_request == MPI_REQUEST_NULL) {
    // prepare for the first receive (non-blocking)
    MPI_Irecv(buffer,max_buffer_size,MPI_BYTE,MPI_ANY_SOURCE,MPI_ANY_TAG,
              comm,&receive_request);
  }
  int mtag;
  MPI_Test(&receive_request,&mtag,MPI_STATUS_IGNORE);
  if (mtag == 1) {
    // deserialize and process the incoming data
    int opr_id = *(const int*)buffer;
    int tag = *(const int*)(buffer+sizeof(int));
    int len = *(const int*)(buffer+2*sizeof(int));
    void* data;
    if (opr_id < 0)
      data = (void*)*(const long*)(buffer+3*sizeof(int));
    else deserialize(data,buffer+3*sizeof(int),len,operations[opr_id]->encoded_type);
    // prepare for the next receive (non-blocking)
    MPI_Irecv(buffer,max_buffer_size,MPI_BYTE,MPI_ANY_SOURCE,
              MPI_ANY_TAG,comm,&receive_request);
    handle_received_message(tag,opr_id,data,len+3*sizeof(int));
    return 1;
  }
  return 0;
}

void kill_receiver () {
  if (receive_request != MPI_REQUEST_NULL) {
    MPI_Cancel(&receive_request);
    MPI_Request_free(&receive_request);
    receive_request = MPI_REQUEST_NULL;
  }
}

void run_receiver () {
  if (receive_request != MPI_REQUEST_NULL)
    kill_receiver();
  while (!stop_receiver) {
    if (check_communication() == 0)
      this_thread::sleep_for(chrono::milliseconds(1));
  }
}

int mpi_abort () {
  return MPI_Abort(comm,0);
}

mutex send_data_mutex;

// send the opr cached data to a remote rank
void send_data ( int rank, void* data, int opr_id, int tag ) {
  // needs to be static and locked
  static char* buffer = new char[max_buffer_size];
  lock_guard<mutex> lock(send_data_mutex);
  // serialize data into a byte array
  Opr* opr = operations[opr_id];
  *(int*)buffer = opr_id;
  *(int*)(buffer+sizeof(int)) = tag;
  int len = serialize(data,buffer+3*sizeof(int),opr->encoded_type);
  *(int*)(buffer+2*sizeof(int)) = len;
  info("    sending %d bytes to %d (opr %d)",
       len+3*sizeof(int),rank,opr_id);
  if (!enable_recovery)
    MPI_Send(buffer,len+3*sizeof(int),MPI_BYTE,rank,tag,comm);
  else {
    MPI_Request sr = MPI_REQUEST_NULL;
    int ierr = MPI_Isend(buffer,len+3*sizeof(int),MPI_BYTE,rank,tag,comm,&sr);
    if (ierr != MPI_SUCCESS)
      mpi_error(rank,ierr);
    int mtag;
    ierr = MPI_Test(&sr,&mtag,MPI_STATUS_IGNORE);
    if (ierr != MPI_SUCCESS)
      mpi_error(rank,ierr);
    int count = 0;
    while (!mtag && count < max_wait_time) {
      count++;
      this_thread::sleep_for(chrono::milliseconds(1));
    }
    if (!mtag && !skip_work) {
      MPI_Cancel(&sr);
      MPI_Request_free(&sr);
      // executor rank is not responding => start recovery
      mpi_error(rank,-1);
    } else MPI_Request_free(&sr);
  }
}

// send one long int to a remote rank
void send_long ( int rank, long value, int tag ) {
  static int* buffer = new int[5];
  buffer[0] = -1;
  buffer[1] = tag;
  buffer[2] = 2;
  *(long*)(buffer+3) = value;
  MPI_Send(buffer,5*sizeof(int),MPI_BYTE,rank,tag,comm);
}

// accumulate values at the coordinator O(n)
bool accumulate ( bool value, MPI_Op acc ) {
  long n = value ? 1L : 0L;
  send_long(coordinator,n,5);
  accumulator.acc = acc;
  accumulator.exit = false;
  accumulator.total = accumulator.acc != MPI_LOR;
  while ( !accumulator.exit )
    this_thread::sleep_for(chrono::milliseconds(1));
  return accumulator.total;
}

// logical and of all b's from all executors (blocking)
bool wait_all ( bool b ) {
  if (inMemory)
    return b;
  if (skip_work)
    return false;
  if (enable_recovery)
    return accumulate(b,MPI_LAND);
  int in[1] = { ( b ? 0 : 1 ) };
  int ret[1];
  MPI_Allreduce(in,ret,1,MPI_INT,MPI_LOR,comm);
  return ret[0] == 0;
}

void mpi_barrier_no_recovery () {
  MPI_Barrier(comm);
}

void mpi_barrier () {
  if (enable_recovery)
    accumulate(false,MPI_LOR); // O(n)
  else MPI_Barrier(comm);      // O(logn)
}

int getCoordinator () { return coordinator; }

bool isCoordinator () {
  return executor_rank == coordinator;
}

void mpi_startup ( int argc, char* argv[], int block_dim_size ) {
  // can fit 5 blocks of double
  max_buffer_size = max(100000LU,5*sizeof(double)*block_dim_size*block_dim_size);
  if (inMemory)
    return;
  // Multiple threads may call MPI, with no restrictions
  int provided;
  MPI_Init_thread(&argc,&argv,MPI_THREAD_MULTIPLE,&provided);
  if (provided != MPI_THREAD_MULTIPLE) {
    printf("Required MPI_THREAD_MULTIPLE to run MPI+OpenMP");
    MPI_Finalize();
    exit(-1);
  }
  MPI_Comm_set_errhandler(comm,MPI_ERRORS_RETURN);
  MPI_Comm_rank(comm,&executor_rank);
  MPI_Comm_size(comm,&num_of_executors);
  coordinator = 0;
  for ( int i = 0; i < num_of_executors; i++ )
    active_executors.push_back(i);
  char machine_name[256];
  gethostname(machine_name,255);
  int local_rank = 0;
  char* lc = getenv("OMPI_COMM_WORLD_LOCAL_RANK");
  if (lc == nullptr)
    lc = getenv("MV2_COMM_WORLD_LOCAL_RANK");
  if (lc != nullptr)
    local_rank = atoi(lc);
  int ts;
  #pragma omp parallel
  { ts = omp_get_num_threads(); }
  printf("Using executor %d: %s/%d (threads: %d)\n",
         executor_rank,machine_name,local_rank,ts);
  reset_accumulator();
}

void mpi_finalize () {
  if (!inMemory)
    MPI_Finalize();
}

File: scala-planner/main/scala/edu/uta/diablo/Communication.scala
/*
 * Copyright © 2024-2024 University of Texas at Arlington
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package edu.uta.diablo

import PlanGenerator.{operations=>_,loadBlocks=>_,_}
import Runtime._
import mpi._
import mpi.MPI._
import java.nio.ByteBuffer
import java.io._


@transient
// MPI communicator using openMPI (one executor per node)
object Executor {
  var executor_rank: WorkerID = 0
  var coordinator: WorkerID = 0
  // the buffer must fit few blocks (one block of doubles is 8MBs)
  val max_buffer_size = 500000000
  val buffer: ByteBuffer = newByteBuffer(max_buffer_size)
  var receive_request: Request = _
  var comm: Intracomm = _//MPI.COMM_WORLD
  var enable_recovery: Boolean = false
  var failed_executors: List[WorkerID] = Nil
  var active_ex  }

  def serialization_error ( ex: IOException ) {
    info("Serialization error: "+ex.getMessage)
    ex.printStackTrace()
    comm.abort(-1)
  }

  def error ( msg: String ) {
    info("*** "+msg)
    comm.abort(-1)
  }

  // process the incoming message
  def handle_received_message ( tag: Byte, opr_id: OprID, data: Any, len: Int ) {
    val opr = operations(opr_id)
    if (tag == 0) {  // cache data and check for ready operations
      info("    received "+(len+4)+" bytes at "+executor_rank+" (opr "+opr_id+")")
      cache_data(opr,data)
      opr.status = completed
      enqueue_ready_operations(opr_id)
    } else if (tag == 2) {   // cache data
      info("    received "+(len+4)+" bytes at "+executor_rank)
      cache_data(opr,data)
      opr.status = completed
    } else if (tag == 3) {  // coordinator will decide who will replace a failed executor
      val failed_executor = data.asInstanceOf[Int]
      if (!failed_executors.contains(failed_executor)) {
        skip_work = true
        failed_executors = failed_executor::failed_executors
        active_executors = active_executors.filter(_ != failed_executor)
        val candidates = active_executors.filter(_ != coordinator)
        val new_executor = candidates((Math.random()*candidates.length).toInt)
        info("The coordinator has decided to replace the failed executor "
             +failed_executor+" with "+new_executor)
        send_data(active_executors,(failed_executor,new_executor),0,4)
      }
    } else if (tag == 4) {  // recovery at all executors (sent from coordinator)
      val (failed_executor,new_executor) = data.asInstanceOf[(Int,Int)]
      if (!failed_executors.contains(failed_executor))
        failed_executors = failed_executor::failed_executors
      active_executors = active_executors.filter(_ != failed_executor)
      Runtime.recover(failed_executor,new_executor)
    } else if (tag == 5) {  // synchronize accumulation (on coordinator)
      accumulator.values = data::accumulator.values
      accumulator.count += 1
      if (!skip_work && accumulator.count == active_executors.length) {
        accumulator.total = accumulator.values.reduce(accumulator.acc)
        accumulator.count = 0
        send_data(active_executors,accumulator.total,6)
        //info("Aggregate "+accumulator.values+" = "+accumulator.total)
      }
    } else if (tag == 6) {   // release the barrier synchronization after accumulation
      accumulator.reset()
      accumulator.total = data
    } else if (tag == 1) {   // partial/final reduction
      opr match {
        case ReduceOpr(s,valuep,fid)
          => // partial reduce ReduceOpr
             info("    received partial reduce result of "+(len+4)+" bytes at "
                  +executor_rank+" (opr "+opr_id+")")
             val op = functions(fid).asInstanceOf[((Any,Any))=>Any]
             if (opr.cached == null) {
               stats.cached_blocks += 1
               stats.max_cached_blocks = Math.max(stats.max_cached_blocks,stats.cached_blocks)
               opr.cached = data
             } else if (valuep)
                      // total aggregation
                      opr.cached = op((opr.cached,data))
               else { val x = opr.cached.asInstanceOf[(Any,Any)]
                      val y = data.asInstanceOf[(Any,Any)]
                      // merge the current state with the incoming partially reduced data
                      opr.cached = (x._1,op((x._2,y._2)))
                    }
             opr.reduced_count -= 1
             if (opr.reduced_count == 0) {
               // completed ReduceOpr
               info("    completed reduce of opr "+opr_id+" at "+executor_rank)
               stats.reduce_operations += 1
               opr.status = computed
               enqueue_ready_operations(opr_id)
             }
        case _ => ;
      }
    } else error("Unknown received tag: "+tag+" at exec "+executor_rank)
  }

  // check if there is a request to send data (array blocks); if there is, get the data and cache it
  def check_communication (): Int = {
    if (receive_request == null)
      try {
        buffer.clear()
        // prepare for the first receive (non-blocking)
        receive_request = comm.iRecv(buffer,buffer.capacity(),BYTE,ANY_SOURCE,ANY_TAG)
      } catch { case ex: MPIException
                  => mpi_error(-1,ex) }
    if (receive_request.test()) {
      // deserialize and cache the incoming data
      val len = buffer.getInt()
      val bb = Array.ofDim[Byte](len)
      buffer.get(bb)
      val bs = new ByteArrayInputStream(bb,0,len)
      val is = new ObjectInputStream(bs)
      val opr_id = is.readInt()
      val tag = is.readByte()
      val data = try { is.readObject()
                     } catch { case ex: IOException
                                 => serialization_error(ex); null }
      is.close()
      try {
        receive_request.free()
        buffer.clear()
        // prepare for the next receive (non-blocking)
        receive_request = comm.iRecv(buffer,buffer.capacity(),BYTE,ANY_SOURCE,ANY_TAG)
      } catch { case ex: MPIException
                  => mpi_error(-1,ex) }
      new Thread() {
        override def run () {
          handle_received_message(tag,opr_id,data,len)
        }
      }.start()
      1
    } else 0
  }

  // send the result of an operation (array blocks) to other executors
  def send_data ( ranks: List[WorkerID], data: Any, opr_id: OprID, tag: Int ) {
    if (data == null)
      error("null data sent for "+opr_id+" to "+ranks.mkString(",")+" from "+executor_rank)
    // serialize data into a byte array
    val bs = new ByteArrayOutputStream(max_buffer_size)
    val os = new ObjectOutputStream(bs)
    try {
      os.writeInt(opr_id)
      os.writeByte(tag)
      os.writeObject(data)
      os.close()
    } catch { case ex: IOException
                => serialization_error(ex) }
    val ba = bs.toByteArray
    val bb = newByteBuffer(ba.length+4)
    bb.putInt(ba.length).put(ba)
    info("    sending "+(ba.length+4)+" bytes from "+executor_rank
         +" to "+ranks.mkString(",")+(if (opr_id <= 0) " (tag "+tag+")" else " (opr "+opr_id+")"))
    for ( rank <- ranks )
      try {
        if (!enable_recovery)
          comm.send(bb,ba.length+4,BYTE,rank,tag)
        else {
          var sr = comm.iSend(bb,ba.length+4,BYTE,rank,tag)
          var count = 0
          while (!sr.test() && count < max_wait_time) {
            count += 1
            Thread.sleep(1)
          }
          if (!sr.test() && !Runtime.skip_work) {
            sr.cancel()
            sr.free()
            throw new MPIException("Executor "+rank+" is not responding")
          } else sr.free()
        }
      } catch { case ex: MPIException
                  => mpi_error(rank,ex) }
  }

  // send data to executors
  def send_data ( ranks: List[WorkerID], data: Any, tag: Int ) {
    // serialize data into a byte array
    val bs = new ByteArrayOutputStream(max_buffer_size)
    val os = new ObjectOutputStream(bs)
    try {
      os.writeInt(0)
      os.writeByte(tag)
      os.writeObject(data)
      os.close()
    } catch { case ex: IOException
                => serialization_error(ex) }
    val ba = bs.toByteArray
    val bb = newByteBuffer(ba.length+4)
    bb.putInt(ba.length).put(ba)
    for ( rank <- ranks )
      try {
        comm.send(bb,ba.length+4,BYTE,rank,tag)
      } catch { case ex: MPIException
                  => mpi_error(rank,ex) }
  }

  // receive the result of an operation (array blocks) from another executor
  def receive_data ( rank: WorkerID ): Any = {
    buffer.clear()
    try {
      comm.recv(buffer,buffer.capacity(),BYTE,ANY_SOURCE,ANY_TAG)
    } catch { case ex: MPIException
                => mpi_error(rank,ex) }
    val len = buffer.getInt()
    val bb = Array.ofDim[Byte](len)
    buffer.get(bb)
    val bs = new ByteArrayInputStream(bb,0,len)
    val is = new ObjectInputStream(bs)
    val opr_id = is.readInt()
    val tag = is.readByte()
    val data = try { is.readObject()
                   } catch { case ex: IOException
                               => serialization_error(ex); null }
    info("    received "+(len+4)+" bytes at "+executor_rank
         +" from "+rank+" (opr "+opr_id+")")
    is.close()
    operations(opr_id).cached = data
    operations(opr_id).status = completed
    data
  }
 
  def broadcast_plan () {
    try {
      if (isCoordinator()) {
        val bs = new ByteArrayOutputStream(10000)
        val os = new ObjectOutputStream(bs)
        os.writeInt(loadBlocks.length)
        os.writeObject(operations)
        os.writeObject(functions)
        os.close()
        val a = bs.toByteArray
        comm.bcast(Array(a.length),1,INT,0)
        comm.bcast(a,a.length,BYTE,0)
        for ( opr_id <- operations.indices ) {
          val x = operations(opr_id)
          x match {
            case LoadOpr(b)
              if x.node > 0
              => send_data(List(x.node),loadBlocks(b),opr_id,0)
            case _ => ;
          }
        }
      } else {
        val len = Array(0)
        comm.bcast(len,1,INT,0)
        val plan_buffer = Array.fill[Byte](len(0))(0)
        comm.bcast(plan_buffer,len(0),BYTE,0)
        val bs = new ByteArrayInputStream(plan_buffer,0,plan_buffer.length)
        val is = new ObjectInputStream(bs)
        val lb_len = is.readInt()
        loadBlocks = Array.ofDim[Any](lb_len)
        operations = is.readObject().asInstanceOf[Array[Opr]]
        functions = is.readObject().asInstanceOf[Array[Nothing=>Any]]
        is.close()
        for ( x <- operations ) {
          x match {
            case LoadOpr(b)
              if x.node == executor_rank
              => loadBlocks(b) = receive_data(0)
            case _ => ;
          }
        }
      }
      comm.barrier()
    } catch { case ex: MPIException
                => mpi_error(-1,ex)
              case ex: IOException
                => serialization_error(ex) }
  }

  def broadcast ( value: Any ): Any = {
    var result = value
    try {
      if (isCoordinator()) {
        val bs = new ByteArrayOutputStream(10000)
        val os = new ObjectOutputStream(bs)
        os.writeObject(value)
        os.flush()
        os.close()
        val a = bs.toByteArray
        comm.bcast(Array(a.length),1,INT,0)
        comm.bcast(a,a.length,BYTE,0)
      } else {
        val len = Array(0)
        comm.bcast(len,1,INT,0)
        val buffer = Array.fill[Byte](len(0))(0)
        comm.bcast(buffer,len(0),BYTE,0)
        val bs = new ByteArrayInputStream(buffer,0,buffer.length)
        val is = new ObjectInputStream(bs)
        result = is.readObject()
        is.close()
      }
      comm.barrier()
    } catch { case ex: MPIException
                => mpi_error(-1,ex)
              case ex: IOException
                => serialization_error(ex) }
    result
  }

  def broadcast_exit_points ( es: List[OprID] ): List[OprID] = {
    exit_points = broadcast(if (isCoordinator()) es else null).asInstanceOf[List[OprID]]
    exit_points
  }

  // accumulate values at the coordinator O(n)
  def accumulate[T] ( value: T, acc: (T,T) => T ): T = {
    send_data(List(coordinator),value,5)
    accumulator.acc = acc.asInstanceOf[(Any,Any)=>Any]
    accumulator.exit = false
    while ( !accumulator.exit )
      Thread.sleep(10)
    accumulator.total.asInstanceOf[T]
  }

  // exit when the queues of all executors are empty and exit points have been computed
  def exit_poll (): Boolean = {
    def and ( x: Boolean, y: Boolean ) = x && y
    if (skip_work)
      false
    else {
      val b = (ready_queue.isEmpty
               && send_queue.isEmpty
               && exit_points.forall {
                     x => val opr = operations(x)
                          opr.node != executor_rank || hasCachedValue(opr) })
      if (enable_recovery) {
        if (!b)
          info("    cannot exit at "+executor_rank+": needs "+exit_points.filter {
                     x => val opr = operations(x)
                          opr.node == executor_rank && !hasCachedValue(opr) })
        accumulate(b,and)
      } else try {
        val in = Array[Byte](if (b) 0 else 1)
        val ret = Array[Byte](0)
        comm.allReduce(in,ret,1,BYTE,LOR)
        ret(0) == 0
      } catch { case ex: MPIException
                  => mpi_error(-1,ex); false }
    }
  }

  def collect_statistics (): Statistics = {
    def plus ( x: Int, y: Int ) = x+y
    def max ( x: Int, y: Int ) = Math.max(x,y)
    if (enable_recovery)
      new Statistics(accumulate(stats.apply_operations,plus),
                     accumulate(stats.reduce_operations,plus),
                     accumulate(stats.cached_blocks,plus),
                     accumulate(stats.max_cached_blocks,max))
    else try {
      val in = Array[Int](stats.apply_operations,
                          stats.reduce_operations,
                          stats.cached_blocks)
      val res = Array.fill[Int](in.length)(0)
      comm.reduce(in,res,in.length,INT,SUM,0)
      val max_in = Array(stats.max_cached_blocks)
      val max = Array(0)
      comm.reduce(max_in,max,1,INT,MAX,0)
      new Statistics(res(0),res(1),res(2),max(0))
    } catch { case ex: MPIException
                => mpi_error(-1,ex); stats }
  }
}


object Communication {
  import Executor._
  var num_of_executors: Int = 1

  def barrier () {
    def or ( x: Boolean, y: Boolean ) = x || y
    if (!enable_recovery)
      comm.barrier()    // O(logn)
    else accumulate(false,or)  // O(n)
  }

  def isCoordinator (): Boolean
    = (executor_rank == coordinator)

  def mpi_startup ( args: Array[String] ) {
    try {
      InitThread(args,THREAD_FUNNELED)
      comm = MPI.COMM_WORLD
      // don't abort on MPI errors
      comm.setErrhandler(ERRORS_RETURN)
      executor_rank = comm.getRank
      num_of_executors = comm.getSize
      active_executors = 0.until(num_of_executors).toList
      val local_rank = System.getenv("OMPI_COMM_WORLD_LOCAL_RANK").toInt
      info("Using executor "+executor_rank+": "+getProcessorName+"/"+local_rank+" with "
           +(java.lang.Runtime.getRuntime.availableProcessors())+" Java threads and "
           +(java.lang.Runtime.getRuntime.totalMemory()/1024/1024/1024)+" GBs")
    } catch { case ex: MPIException
                => info("MPI error: "+ex.getMessage)
                   ex.printStackTrace()
                   comm.abort(-1)
            }
  }

  def mpi_finalize () {
    info("Executor "+executor_rank+" is exiting")
    if (!enable_recovery)
      Finalize()
    else System.exit(0)
  }
